FEATURE: Assets sources support for Neos Media

robertlemke opened the pull request:
This change introduces low-level support for third-party "asset sources" which allow for seamless integration of external asset management services.

The intention of this change is to establish a beta version of the API which can be used for developing first asset source plugins. The newly introduced API might change slightly during that process and thus should not be considered as stable. The native Neos asset source, which is part of this change, should work like it did before the introduction of this new feature.

robertlemke commented:
Just pushed a new version which should – after a `doctrine:migrate` just work like before, but internally using Asset Sources. The only actual feature currently missing is the asynchronous generation of thumbnails (to be done before merge). But nothing should be broken, everything should work fine.

Therefore, a first review would be appreciated - and if it's only trying out the media functionality to check if everything still works.

Things to try:
- upload assets
- import sites
- pick assets for a content element
- replace existing assets
- tag and untag assets
- put assets into collections
- thumbnail view versus list view
- detail view of an asset (edit / show)

robertlemke commented:
Pushed the re-implemented support for asynchronous thumbnails. Now missing: Postgresql Migration and documentation.

robertlemke commented:
Regarding the failed Travis test for the YAML configuration: it's a valid YAML syntax which works in the Symfony YAML parser I'm using here locally, so dependencies / versions are something we should look into.

kitsunet commented on Neos.Media.Browser/Classes/Controller/AssetController.php:
```diff
@@ -854,9 +799,158 @@ public function addFlashMessage($messageBody, $messageTitle = '', $severity = Me
+        $assetSourceIdentifier = $this->browserState->getActiveAssetSourceIdentifier();
+        if (!isset($this->assetSources[$assetSourceIdentifier])) {
+            $assetSourceIdentifiers = array_keys($this->assetSources);
+            $assetSourceIdentifier = reset($assetSourceIdentifiers);
```
Don't we want to error here if there was an `$assetSourceIdentifier` set that was not found?

robertlemke replied:
I'd say, rather not show an error. Because it could be that asset source have been reconfigured and you still have an old identifier in your session - then falling back to another asset source would be okay, I guess.

kitsunet commented on Neos.Media.Browser/Classes/Controller/AssetProxyController.php:
```diff
@@ -0,0 +1,168 @@
+     */
+    public function initializeObject()
+    {
+        foreach ($this->assetSourcesConfiguration as $assetSourceIdentifier => $assetSourceConfiguration) {
```
This stuff could be DRY'd and extracted from here and the other controller.

robertlemke replied:
"Three strikes and you refactor" is the rule I usually apply.

kitsunet commented on Neos.Media.Browser/Classes/Controller/AssetController.php:
```diff
@@ -379,22 +316,89 @@ public function replaceAssetResourceAction(Asset $asset)
+            throw new \RuntimeException('Given asset source is not configured.', 1509702178632);
+        }
+
+        $assetBrowser = $this->assetSources[$assetSourceIdentifier]->getAssetProxyRepository();
```
In all other places this variable is called `$assetProxyRepository`

robertlemke replied:
good catch, that's a leftover from the old code.

kitsunet commented on Neos.Media.Browser/Classes/Controller/AssetController.php:
```diff
@@ -379,22 +316,89 @@ public function replaceAssetResourceAction(Asset $asset)
+
+            $tags = [];
+            if ($assetProxyRepository instanceof SupportsTagging) {
+                if ($assetProxyRepository instanceof SupportsCollections) {
```
Seems this inner if could be added to the condition of the one above.

robertlemke replied:
✅

kitsunet commented on Neos.Media.Browser/Classes/Controller/ImageController.php:
```diff
@@ -28,14 +30,31 @@ class ImageController extends AssetController
-            $asset = $asset->getOriginalAsset();
+        if ($assetSourceIdentifier !== null && $assetProxyIdentifier !== null) {
+            parent::editAction($assetSourceIdentifier, $assetProxyIdentifier);
+            return;
```
Shouldn't this actually `return parent::editAction($assetSourceIdentifier, $assetProxyIdentifier);` ?

robertlemke replied:
The parent `editAction` is declared as `@return void`, so if it was `return parent::editAction()…`, Phpstorm complains …

But right, it looks a bit suspicious.

kitsunet commented on Neos.Media.Browser/Classes/Controller/ImageController.php:
```diff
@@ -28,14 +30,31 @@ class ImageController extends AssetController
+            /** @var ImportedAsset $importedAsset */
+            $importedAsset = $this->importedAssetRepository->findOneByLocalAssetIdentifier($asset->getIdentifier());
+            parent::editAction($asset->getAssetSourceIdentifier(), $importedAsset ? $importedAsset->getRemoteAssetIdentifier() : $asset->getIdentifier());
+            return;
```
See above.

kitsunet commented on Neos.Media.Browser/Classes/Service/FileTypeIconService.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Service that retrieves an icon for the file type of a given filename
+ */
+class FileTypeIconService
```
Doesn't this exist in the Media package?

robertlemke replied:
Yes, in a slightly different way (it needs Asset objects instead of a filename). I'll move this to the Media package and just the code accordingly.

kitsunet commented on Neos.Media.Browser/Classes/ViewHelpers/FileTypeIconViewHelper.php:
```diff
@@ -0,0 +1,77 @@
+ * <img src="_Resources/Static/.../Icons/16px/jpg.png" height="16" alt="file type alt text" />
+ * </output>
+ */
+class FileTypeIconViewHelper extends AbstractTagBasedViewHelper
```
Doesn't this exist in the Media package?

robertlemke replied:
Yes, will be consolidated.

kitsunet commented on Neos.Media.Browser/Classes/ViewHelpers/ThumbnailViewHelper.php:
```diff
@@ -0,0 +1,78 @@
+ * <img src="https://my-asset-management.com/thumbnails/espresso.jpg" width="120" height="180" alt="a sample asset without scaling" />
+ * </output>
+ */
+class ThumbnailViewHelper extends AbstractTagBasedViewHelper
```
Doesn't this exist in the Media package?

robertlemke replied:
Yes, will but this one is different. I'll add a comment to declare it as "internal" as I don't want to add this functionality too early into the official / public thumbnail view helper.

kitsunet commented on Neos.Media.Browser/composer.json:
```diff
@@ -1,91 +1,25 @@
+            "package-key": "Neos.Media.Browser"
+        },
         "applied-flow-migrations": [
-            "TYPO3.TYPO3CR-20150510103823",
```
Don't we need those?

kitsunet commented on Neos.Media.Browser/composer.json:
```diff
@@ -1,91 +1,25 @@
-        "doctrine/common": "~2.7.0",
-        "doctrine/orm": "~2.5.0"
+        "php": "^7.1",
+        "neos/neos": "3.*",
```
Should be self.version as before

robertlemke replied:
That slipped in from my fork. 🤭

kitsunet commented on Neos.Media.Browser/composer.json:
```diff
@@ -1,91 +1,25 @@
-        "doctrine/orm": "~2.5.0"
+        "php": "^7.1",
+        "neos/neos": "3.*",
+        "neos/media-browser": "3.*"
```
same

kitsunet commented on Neos.Media/Classes/Domain/Model/Asset.php:
```diff
@@ -144,6 +171,12 @@ public function initializeObject($initializationCause)
             $this->emitAssetCreated($this);
         }
+
+        foreach ($this->assetSourcesConfiguration as $assetSourceIdentifier => $assetSourceConfiguration) {
```
again could be DRY'd.

kitsunet commented on Neos.Media/Classes/Domain/Model/Asset.php:
```diff
@@ -412,6 +445,56 @@ public function setAssetCollections(Collection $assetCollections)
+    /**
+     * @param string $assetSourceIdentifier
+     */
+    public function setAssetSourceIdentifier(string $assetSourceIdentifier): void
```
Should this actually be settable? What happens if I set this to something else?

robertlemke replied:
Yes, needs to be settable, but I'll document it as internal-use only.

kitsunet commented on Neos.Media/Classes/Domain/Model/Asset.php:
```diff
@@ -412,6 +445,56 @@ public function setAssetCollections(Collection $assetCollections)
+        return $this->assetSourceIdentifier;
+    }
+
+    public function getAssetSource(): ?AssetSource
```
Does the Asset actually need this? And shouldn't the resource be instantiated lazily in here?

robertlemke replied:
No, that's not needed (anymore), I'll refactor this, also to be initialized lazily.

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetProxy/AssetProxy.php:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Interface for a stand-in object of remote or already imported assets from an asset source.
+ */
+interface AssetProxy
```
Currently our CGL say to add `Interface` to all Interface names. A change to that policy was rebuked a while ago.

robertlemke replied:
will fix

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetProxy/HasRemoteOriginal.php:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Interface for an Asset Proxy which depicts an asset whose original binary data is stored in a remote location
+ */
+interface HasRemoteOriginal
```
See above

robertlemke replied:
will fix

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetProxy/SupportsIptcMetadata.php:
```diff
@@ -0,0 +1,46 @@
+ * Note that the property names are UpperCamelCase and as defined in the IPTC IIM specification.
+ * Examples: "Title", "Keywords", "CopyrightNotice"
+ */
+interface SupportsIptcMetadata
```
See above

kitsunet replied:
Shouldn't there be a separate `IptcMetadataInterface` and the respective Asset just have a `getIptcMetadata`?

robertlemke replied:
nah, I'd rather use adjectives still, where possible.

kitsunet replied:
No, what I meant is, shouldn't metadata be a separate object, so the asset can only give you the metadata object and those methods are actually in that separate object?

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetProxyQuery.php:
```diff
@@ -0,0 +1,55 @@
+ * source code.
+ */
+
+interface AssetProxyQuery
```
See above.

robertlemke replied:
will fix

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetProxyQueryResult.php:
```diff
@@ -0,0 +1,38 @@
+
+use Neos\Media\Domain\Model\AssetSource\AssetProxy\AssetProxy;
+
+interface AssetProxyQueryResult extends \Countable, \Iterator, \ArrayAccess
```
See above

robertlemke replied:
will fix

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetProxyRepository.php:
```diff
@@ -0,0 +1,61 @@
+use Neos\Media\Domain\Model\AssetSource\AssetProxy\AssetProxy;
+use Neos\Media\Domain\Model\Tag;
+
+interface AssetProxyRepository
```
See above

robertlemke replied:
will fix

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetSource.php:
```diff
@@ -0,0 +1,48 @@
+ * source code.
+ */
+
+interface AssetSource
```
see above

robertlemke replied:
will fix

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxy.php:
```diff
@@ -0,0 +1,189 @@
+    public function __construct(AssetInterface $asset, NeosAssetSource $assetSource)
+    {
+        if (!$asset instanceof Asset) {
+            throw new \RuntimeException(sprintf('%s currently only support the concrete Asset implementation because several methods are not part of the AssetInterface yet.', __CLASS__), 1509635540408);
```
Why don't we make it part of the interface?

robertlemke replied:
That would be a bigger breaking change. We should tackle this in a separate PR, to give it the required attention. I mean, sure, this would be a good time to do (for a major release). But time …

kitsunet commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetSource.php:
```diff
@@ -0,0 +1,184 @@
+                        if (!is_bool($optionValue)) {
+                            throw new \InvalidArgumentException(sprintf('Asset source option "%s" specified for Neos asset source "%s" must be either true or false. Please check your settings.', $optionName, $assetSourceIdentifier), 1522927471208);
+                        }
+                        $this->asyncThumbnails = (bool)$optionValue;
```
If we really need to be so strict as to chedck for boolean above, we don't really need to cast here...

kitsunet commented on Neos.Media/Classes/Domain/Model/ImportedAsset.php:
```diff
@@ -0,0 +1,152 @@
+        string $assetSourceIdentifier,
+        string $remoteAssetIdentifier,
+        string $localAssetIdentifier,
+        string $localOriginalAssetIdentifier = null,
```
optional argument before required argument!

robertlemke replied:
✅

kitsunet commented on Neos.Media/Classes/Domain/Model/ImportedAsset.php:
```diff
@@ -0,0 +1,152 @@
+    /**
+     * @param string $assetSourceIdentifier
+     */
+    public function setAssetSourceIdentifier(string $assetSourceIdentifier): void
```
Shouldn't we actually have no setters for the properties? As this entity should be immutable.

kitsunet commented on Neos.Media/Classes/Domain/Model/ImportedAssetManager.php:
```diff
@@ -0,0 +1,103 @@
+            return;
+        }
+
+        $variantAsset = $asset;
```
Why not call the argument `$variantAsset`?

kitsunet commented on Neos.Media/Classes/Domain/Model/ImportedAssetManager.php:
```diff
@@ -0,0 +1,103 @@
+        $originalAssetIdentifier = $this->persistenceManager->getIdentifierByObject($originalAsset);
+
+        $originalImportedAsset = $this->importedAssetRepository->findOneByLocalAssetIdentifier($originalAssetIdentifier);
+        if ($originalImportedAsset instanceof ImportedAsset && $variantAsset instanceof AssetSourceAwareInterface) {
```
This check should probably happen earlier to avoid unnecessary queries.

kitsunet commented:
Existing assets need a assetsourceidentifier set in the migration.

robertlemke commented on Neos.Media.Browser/Classes/Controller/AssetController.php:
```diff
@@ -5,58 +5,58 @@
  *
  * (c) Contributors of the Neos Project - www.neos.io
- *
+  *
```
✅

robertlemke commented on Neos.Media.Browser/Classes/Controller/AssetController.php:
```diff
@@ -379,22 +316,89 @@ public function replaceAssetResourceAction(Asset $asset)
+            $assetProxy = $assetBrowser->getAssetProxy($assetProxyIdentifier);
+
+            $this->view->assignMultiple([
+//            'tags' => $assetProxy->getAssetCollections()->count() > 0 ? $this->tagRepository->findByAssetCollections($assetProxy->getAssetCollections()->toArray()) : $this->tagRepository->findAll(),
```
✅

robertlemke commented on Neos.Media.Browser/Classes/Controller/AssetController.php:
```diff
@@ -690,93 +699,29 @@ public function relatedNodesAction(AssetInterface $asset)
-            $relatedNodes[$site->getNodeName()]['nodes'][] = [
-                'node' => $node,
-                'documentNode' => $documentNode
+            $documentNodeIdentifier = $usage->getDocumentNode() instanceof NodeInterface ? $usage->getDocumentNode()->getIdentifier() : null;
```
Actually, the whole (rather new) part handling inaccessible usages etc. is missing. Probably because I started my fork before that was implemented.

I will re-add that functionality and push it to this PR.

robertlemke commented:
@kitsunet thanks a lot for your review so far! It tried to tackle everything you mentioned and just pushed a couple of new commits. Best is, if you go through them commit-by-commit.

The most changes addressing the issues you mentioned in your comments, are contained in one bigger commit though.

dfeyer commented:
Will try a full review tomorrow in the evening.

dfeyer commented:
@robertlemke Did you have any documentation around this feature ? Even if it's in beta, it can help

robertlemke commented:
@dfeyer I can publish my Google Drive prototype tomorrow, I guess that helps!

daniellienert commented:
Testplan:
- [x] upload assets
- [x] import sites
- [x] pick assets for a content element
- [x] replace existing assets
- [x] tag and untag assets
- [x] put assets into collections
- [x] thumbnail view versus list view
- [x] detail view of an asset (edit / show)

daniellienert commented:
@robertlemke: With this change, all asset variants are shown in the media module:

Master:
![image](https://user-images.githubusercontent.com/642226/38954693-a341aa6e-4352-11e8-911e-b7dd5e775d30.png)

With this change:
![image](https://user-images.githubusercontent.com/642226/38954666-9224c90a-4352-11e8-92a3-42bd327c1101.png)

daniellienert commented on Neos.Media.Browser/Classes/ViewHelpers/Controller/PaginateController.php:
```diff
@@ -0,0 +1,167 @@
+<?php
+
```
Nitpick: No newline before use statement

daniellienert commented on Neos.Media.Browser/Classes/ViewHelpers/Controller/PaginateController.php:
```diff
@@ -0,0 +1,167 @@
+    /**
+     * @var array
+     */
+    protected $configuration = array('itemsPerPage' => 10, 'insertAbove' => false, 'insertBelow' => true, 'maximumNumberOfLinks' => 99);
```
Nitpick: Do not use legacy array syntax

daniellienert commented on Neos.Media.Browser/Classes/ViewHelpers/Controller/PaginateController.php:
```diff
@@ -0,0 +1,167 @@
+    protected $maximumNumberOfLinks = 99;
+
+    /**
+     * @var integer
```
Should'nt the annotation be just 'int'

daniellienert commented on Neos.Media.Browser/Resources/Private/Translations/da/Main.xlf:
```diff
@@ -68,6 +68,9 @@
       <trans-unit id="editAsset" xml:space="preserve" approved="yes">
 				<source>Edit asset</source>
 			<target xml:lang="da">Rediger fil</target><alt-trans><target xml:lang="da">Rediger</target></alt-trans></trans-unit>
+      <trans-unit id="replaceAssetResource" xml:space="preserve" approved="yes">
```
All these files are managed by crowdin, or am I wrong? So they shouldn't be altered manually.

kdambekalns commented:
That test plan misses showing of asset usages…

> Am 18.04.2018 um 21:32 schrieb Daniel Lienert <notifications@github.com>:
> 
> Testplan:
> 
>  upload assets
>  import sites
>  pick assets for a content element
>  replace existing assets
>  tag and untag assets
>  put assets into collections
>  thumbnail view versus list view
>  detail view of an asset (edit / show)
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetNotFoundException.php:
```diff
@@ -0,0 +1,25 @@
+ * source code.
+ */
+
+use Neos\Flow\Exception;
```
Hey wasn't one of my learnings from the Con, that introducing Neos\Flow\Exception was a bad idea and that one should not use it :) Which talk was ist @kitsunet / @kdambekalns?

robertlemke replied:
I extend the exception because then I have a proper HTTP status code (in this case 404). Would be better if Flow provided an interface for that, but currently it doesn't.

I didn't know that it was bad practice …

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetTypeFilter.php:
```diff
@@ -0,0 +1,57 @@
+    public function __construct(string $assetType)
+    {
+        if (!in_array($assetType, ['All', 'Image', 'Document', 'Video', 'Audio'])) {
+            throw new \InvalidArgumentException('Invalid asset type.', 1511797661398);
```
Would be nice to repeat the wrong type in the exception

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/AssetTypeFilter.php:
```diff
@@ -0,0 +1,57 @@
+    public function __construct(string $assetType)
+    {
+        if (!in_array($assetType, ['All', 'Image', 'Document', 'Video', 'Audio'])) {
+            throw new \InvalidArgumentException('Invalid asset type.', 1511797661398);
```
1511797661398 is too long for a timestamp / includes millis

robertlemke replied:
That's taken from some existing code, but I'll fix it.

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxyInterface.php:
```diff
@@ -0,0 +1,189 @@
+    public function __construct(AssetInterface $asset, NeosAssetSource $assetSource)
+    {
+        if (!$asset instanceof Asset) {
+            throw new \RuntimeException(sprintf('%s currently only support the concrete Asset implementation because several methods are not part of the AssetInterface yet.', __CLASS__), 1509635540408);
```
1511797661398 is too long for a timestamp / includes millis

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxyRepositoryInterface.php:
```diff
@@ -0,0 +1,266 @@
+    {
+        $query = $this->filterOutImportedAssetsFromOtherAssetSources($this->assetRepository->createQuery());
+        $query = $this->filterOutAssetsFromOtherAssetCollections($query);
+        return $query->count();
```
Shouldn't this also use $query = $this->filterOutImageVariants($query);

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxyRepositoryInterface.php:
```diff
@@ -0,0 +1,266 @@
+    {
+        $queryResult = $this->assetRepository->findUntagged($this->activeAssetCollection);
+        $query = $this->filterOutImportedAssetsFromOtherAssetSources($queryResult->getQuery());
+        return new NeosAssetProxyQueryResultInterface($query->execute(), $this->assetSource);
```
Shouldn't this also use $query = $this->filterOutImageVariants($query);

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxyRepositoryInterface.php:
```diff
@@ -0,0 +1,266 @@
+    {
+        $queryResult = $this->assetRepository->findBySearchTermOrTags($searchTerm, [], $this->activeAssetCollection);
+        $query = $this->filterOutImportedAssetsFromOtherAssetSources($queryResult->getQuery());
+        return new NeosAssetProxyQueryResultInterface($query->execute(), $this->assetSource);
```
Shouldn't this also use $query = $this->filterOutImageVariants($query);

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxyRepositoryInterface.php:
```diff
@@ -0,0 +1,266 @@
+    {
+        $queryResult = $this->assetRepository->findAll($this->activeAssetCollection);
+        $query = $this->filterOutImportedAssetsFromOtherAssetSources($queryResult->getQuery());
+        return new NeosAssetProxyQueryResultInterface($query->execute(), $this->assetSource);
```
Shouldn't this also use $query = $this->filterOutImageVariants($query);

mgoldbeck commented:
@robertlemke I merged master with the merged #2009 into your Pull Request so Detail View of asset is shown now.

![detail](https://user-images.githubusercontent.com/1336044/38957655-74990c3a-435b-11e8-857d-f0154e4a5cf5.jpg)

* Image Variants are shown in the Media Modul
![variants](https://user-images.githubusercontent.com/1336044/38957672-79e9343a-435b-11e8-86ae-3eb6a6913bbe.jpg)


* Collection Counter does not work
![variants2](https://user-images.githubusercontent.com/1336044/38957694-85cc8d2e-435b-11e8-8680-bd02ab4f480b.jpg)

Testplan:

- [ ] upload assets
- [ ] import sites
- [ ] pick assets for a content element
- [ ] replace existing assets
- [ ] tag and untag assets
- [ ] put assets into collections
- [ ] thumbnail view versus list view
- [ ] detail view of an asset (edit / show)
- [ ] sorting by last modified and title works
- [ ] the image variants are not shown

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxyRepositoryInterface.php:
```diff
@@ -0,0 +1,266 @@
+use Neos\Media\Domain\Repository\ImageRepository;
+use Neos\Media\Domain\Repository\VideoRepository;
+
+final class NeosAssetProxyRepositoryInterface implements AssetProxyRepositoryInterface, SupportsSorting, SupportsCollections, SupportsTagging
```
Was it really intended to be named `NeosAssetProxyRepositoryInterface`? Shouldn't that rather be `NeosAssetProxyRepository` or even `NeosAssetRepository` ?

daniellienert replied:
Mhm, no?

robertlemke replied:
No, that was of course not the intention ;-) And I fixed that in https://github.com/neos/neos-development-collection/pull/1979/commits/71138fcef84241647419c66ee0ec0fcee1bfe6d5 and https://github.com/neos/neos-development-collection/pull/1979/commits/f0b17f398357feb00a71067be81d714c929d45e5.

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetSource.php:
```diff
@@ -0,0 +1,184 @@
+    public function __construct(string $assetSourceIdentifier, array $assetSourceOptions)
+    {
+        if (preg_match('/^[a-z][a-z0-9-]{0,62}[a-z]$/', $assetSourceIdentifier) !== 1) {
+            throw new \InvalidArgumentException(sprintf('Invalid asset source identifier "%s". The identifier must match /^[a-z][a-z0-9-]{0,62}[a-z]$/', $assetSourceIdentifier), 1513329665386);
```
Non timestamp error code again - I've got a solution to generate timestamps for your PHPStorm if you are interested ;)

robertlemke replied:
Thanks ;-) Actually I use Phpstorm Live Templates for that:
![image](https://user-images.githubusercontent.com/95582/38977264-25f7de4a-43b4-11e8-8197-a1d0802dc257.png)
As far as I can tell, the wrong timestamps are from old code I refactored.

robertlemke replied:
Okay, you're right, actually the Phpstorm `timestamp()` function produces millisecond-based timestamps. And I never spotted the difference! I now switched to a plugin instead which produces second-based timestamps. Thanks, eagle-eyed Daniel!

daniellienert replied:
![image](https://user-images.githubusercontent.com/642226/38997053-ce295c5a-43ec-11e8-8bd1-ac5c2a474e73.png)

daniellienert commented on Neos.Media/Classes/Domain/Model/AssetSource/Neos/NeosAssetProxyInterface.php:
```diff
@@ -0,0 +1,189 @@
+    private $asset;
+
+    /**
+     * @Inject()
```
Mhm, interesting variant to only use the Inject annotation. But inconsistent to the other classes, less expressive and how we show it in the documentations. Are there any benefits for that?

robertlemke replied:
I like this one better and started using it in other projects, but consistency rules!

kitsunet replied:
How can this even work?

robertlemke replied:
@kitsunet well, it does work, that's what the Doctrine annotation driver provides. Actually, importing a whole (sub-) namespace, like we do with "Flow" is sort of a exotic case.

daniellienert commented on Neos.Media/Classes/Domain/Model/ImportedAssetManager.php:
```diff
@@ -0,0 +1,102 @@
+            return;
+        }
+
+        $variantAsset = $asset;
```
I think the renaming is unneeded here, or if its for making things clearer, it should be used in the next line

robertlemke commented:
Thanks y'all for taking the time to review this feature, I really, really appreciate it!

Here's a check list I distilled from your comments. I'll go through these issues and double-check the features again:

- [ ] Bug: Sorting by last modified works while sorting by title returns an empty list
- [ ] Bug: Asset variants are shown in media module
- [ ] Bug: Too long timestamps for exception codes
- [ ] Bug: Shouldn't this also use $query = $this->filterOutImageVariants($query); (NeosAssetProxyRepository)
- [ ] Bug: Collection Counter does not work
- [ ] Style: No newline before use statement (PaginateController)
- [ ] Style: Do not use legacy array syntax (PaginateController)
- [ ] Style: Should'nt the annotation be just 'int' (PaginateController)
- [ ] Style: Inconsistent @Inject() annotation
- [ ] Task: Would be nice to repeat the wrong type in the exception (AssetTypeFilter)
- [ ] Task: Was it really intended to be named NeosAssetProxyRepositoryInterface?

robertlemke commented:
I pushed fixes for all reported issues – will medium-gladly accept suggestions for new bugs ;-)

kitsunet commented:
It seems this is sufficiently reviewed and iterated on, to at least merge it to master and then see if more bugs are found until release? I would merge it tomorrow finally, if no objections.

kitsunet commented:
@robertlemke this is still open https://github.com/neos/neos-development-collection/pull/1979#pullrequestreview-113459087

mgoldbeck commented:
@robertlemke I updated my list but I found also these two ones

- [ ] Sorting the assets by name redirects to Neos Setup
- [ ] referencing an image to an image element from the list view does not work

robertlemke commented:
@daniellienert @kitsunet some background information to why the repository class had a `…RepositoryInterface` name:

When I refactored the interfaces which didn't have an "Interface" suffix to names with "Interface" (for example `AssetProxyRepositoryInterface`), apparently I – or even Phpstorm's refactoring feature – renamed the implementation classes as well. So that wasn't on purpose …

kitsunet commented:
beautiful, I guess we are close to a merge. Travis fails also on master, so let's see that master is green again emrge it into this and then merge this I guess?

robertlemke commented:
Okay, sorry about that late commit, but I pushed another little improvement which improves error handling for connection problems with the asset source. The change is basically

- turn the `AssetSourceConnectionException` into an interface which can be implemented by asset source connection related exceptions
- wrapping relevant parts in `AssetController` into a `try…catch` which will catch connection exceptions
- adjust the Fluid template

It looks like this:

<img width="1394" alt="assetsourceconnectionexception" src="https://user-images.githubusercontent.com/95582/39036894-392f97b6-447f-11e8-9114-682c3e1c6ce1.png">

kdambekalns commented:
So, for me locally all tests (unit, functional behat) pass as of now. 🎉

kdambekalns commented:
@daniellienert and @mgoldbeck should review their approval soonishly… :)

robertlemke commented:
Pushed a commit with code style fixes and a commit which fixes the pick-image-from-list-view problem @mgoldbeck mentioned. I can't reproduce the redirect-to-setup-tool issue though.

Right now I can't spot any further bugs or issues, so from my part … let's merge this chunk!

kitsunet commented:
done, weeeeeeeee, that was the last from the board. Now it's testing master and fixing stuffs

dfeyer commented:
Late in the game, but I spend evening and part of the morning on testing this one, and try to integrate our video storage with the new API, work nicely ;) @robertlemke Kudos, good job. I think I will have a few question, and try to write some documentation or a blog post somewhere when our integration is done.

robertlemke commented:
@dfeyer thanks! I definitely plan to write a little series of blog posts as well, and publish the Google Drive adaptor with it. So if you don't mind, let me prepare that first …

Anyway, there are lots of ideas and plans ahead of course, and the related API will certainly change again. I'm just glad that we have the foundation merged now.
